### PR TITLE
回傳前端Exception Message格式化

### LINKED ISSUE #119: Global Exception Handler Response Format
參考資料之一
https://www.baeldung.com/rest-api-error-handling-best-practices

除了設計格式，還有重點思考
回傳 4xx & 5xx 的條件，甚麼時候該回傳 4xx 甚麼時候該回傳 5xx?
希望能有「具體」的規則可循。

### PR DESCRIPTION
**Pull request by m1a2st**
## Why need this change? / Root cause: 
- the PlatformViewModel need to have a spec 
## Changes made:
- Refactor PlatformException and Exception Handler
## Test Scope / Change impact:
- Exception Message need have a title and message
## Issue
- resolve #119

### CODE REVIEW

File: spring/src/main/kotlin/tw/waterballsa/gaas/spring/aspects/PlatformExceptionHandler.kt [hunk 1]
```diff
@@ -1,21 +1,42 @@
...
+            val requestAttributes = getRequestAttributes() as ServletRequestAttributes
+            return PlatformViewModel(
+                timestamp = LocalDateTime.now().toString(),
+                httpMethod = HttpMethod.resolve(requestAttributes.request.method),
+                path = requestAttributes.request.requestURI,
+                title = exception.title,
+                message = exception.message!!
+            )
+        }
+    }
```
Comment (Wally5077) [inline spring/src/main/kotlin/tw/waterballsa/gaas/spring/aspects/PlatformExceptionHandler.kt:41]: 變成 PlatformException 的擴充方法即可

File: domain/src/main/kotlin/tw/waterballsa/gaas/exceptions/NotFoundException.kt [hunk 2]
```diff
@@ -3,25 +3,31 @@ package tw.waterballsa.gaas.exceptions
...
+            fun identifyBy(identifierName: String, id: Any) = apply {
+                this.message = "Resource (${resourceName.capitalize()}) not found ($identifierName = $id)."
+            }
+
+            fun id(id: Any) = identifyBy("id", id)
+
+            fun title(title: String) = apply { this.title = title }
 
-            fun id(id: Any): NotFoundException = NotFoundException(id, resourceName)
+            fun toNotFoundMessage() = apply { this.message = "${resourceName.capitalize()} not found" }
```
Comment (Wally5077) [inline domain/src/main/kotlin/tw/waterballsa/gaas/exceptions/NotFoundException.kt:28]: Ｑ：為什麼需要調整成這樣？

**Review by Wally5077 — CHANGES_REQUESTED** [review]
**Review done.**

File: domain/src/main/kotlin/tw/waterballsa/gaas/exceptions/NotFoundException.kt (hunk 2 above)
Comment (m1a2st) [reply domain/src/main/kotlin/tw/waterballsa/gaas/exceptions/NotFoundException.kt:29]: A ： 希望此Builder具有一定靈活度

File: domain/src/main/kotlin/tw/waterballsa/gaas/exceptions/enums/ErrorType.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    fun throwingNotFound(message: String): Nothing =
+        notFound(message)
+            .errorType(this)
+            .throwing()
+
+    fun buildNotFoundMessage(message: String): NotFoundException.Companion.NotFoundExceptionBuilder =
+        notFound(message).errorType(this)
+
+    fun <T : Any> buildNotFoundMessage(resourceType: KClass<T>): NotFoundException.Companion.NotFoundExceptionBuilder =
+        notFound(resourceType).errorType(this)
```
Comment (Wally5077) [inline domain/src/main/kotlin/tw/waterballsa/gaas/exceptions/enums/ErrorType.kt:49]: 用原來的方式寫

**Review by Wally5077 — CHANGES_REQUESTED** [review]
**Review done.**

**Review by frankvicky — APPROVED** [review]

File: domain/src/main/kotlin/tw/waterballsa/gaas/domain/User.kt [hunk 4]
```diff
@@ -23,11 +24,11 @@ class User(
         val nicknameByteSize = nickname.toByteArray().size
 
         if (nicknameByteSize < NICKNAME_MINIMUM_BYTE_SIZE) {
-            throw PlatformException("invalid nickname: too short")
+            throw PlatformException(PlatformError.USER_INPUT_INVALID, "invalid nickname: too short")
         }
 
         if (nicknameByteSize > NICKNAME_MAXIMUM_BYTE_SIZE) {
-            throw PlatformException("invalid nickname: too long")
+            throw PlatformException(PlatformError.USER_INPUT_INVALID, "invalid nickname: too long")
```
Comment (Wally5077) [inline domain/src/main/kotlin/tw/waterballsa/gaas/domain/User.kt:31]: **static import**

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/UpdateUserUseCase.kt [hunk 5]
```diff
@@ -28,13 +29,14 @@ class UpdateUserUseCase(
 
     private fun validateNicknameDuplicated(nickname: String) {
         if (userRepository.existsUserByNickname(nickname)) {
-            throw PlatformException("invalid nickname: duplicated")
+            throw PlatformException(PlatformError.USER_INPUT_INVALID, "invalid nickname: duplicated")
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/UpdateUserUseCase.kt:32]: **static import**

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/RegisterGameUsecase.kt [hunk 6]
```diff
@@ -16,7 +17,11 @@ class RegisterGameUsecase(
         gameRegistrationRepository.run {
             val uniqueName = request.uniqueName
             val gameRegistration = when {
-                existsByUniqueName(uniqueName) -> throw GameAlreadyExistsException(uniqueName)
+                existsByUniqueName(uniqueName) -> throw PlatformException(
+                    PlatformError.GAME_EXISTS,
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/RegisterGameUsecase.kt:21]: **static import**

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/LeaveRoomUsecase.kt [hunk 7]
```diff
@@ -23,7 +24,7 @@ class LeaveRoomUsecase(
 
     private fun findRoomById(roomId: Room.Id) =
         roomRepository.findById(roomId)
-            ?: throw notFound(Room::class).id(roomId)
+            ?: throw notFound(PlatformError.ROOM_NOT_FOUND, Room::class).id(roomId)
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/LeaveRoomUsecase.kt:27]: **static import**

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/KickPlayerUsecase.kt [hunk 8]
```diff
@@ -11,7 +12,10 @@ class KickPlayerUsecase(
 ) {
     fun execute(request: Request) {
         with(request) {
-            val room = roomRepository.findById(Room.Id(roomId)) ?: throw notFound(Room::class).message()
+            val room = roomRepository.findById(Room.Id(roomId)) ?: throw notFound(
+                PlatformError.ROOM_NOT_FOUND,
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/KickPlayerUsecase.kt:16]: **static import**

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt [hunk 9]
```diff
@@ -31,23 +32,29 @@ class JoinRoomUsecase(
     private fun validateUserJoinedRoom(userId: String) {
         val hasJoined = roomRepository.hasPlayerJoinedRoom(User.Id(userId))
         if (hasJoined) {
-            throw PlatformException("Player($userId) has joined another room.")
+            throw PlatformException(
+                PlatformError.PLAYER_JOIN_ROOM_ERROR,
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/JoinRoomUsecase.kt:36]: 全檔案 PlatformError **static import**

File: application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/GetUserUseCase.kt [hunk 10]
```diff
@@ -12,7 +13,8 @@ class GetUserUseCase(
     fun execute(request: Request, presenter: Presenter) {
         with(request) {
             val user = userRepository.findByEmail(email)
-                ?: throw notFound(User::class).identifyBy("email", email)
+                ?: throw notFound(PlatformError.PLAYER_NOT_FOUND, User::class)
```
Comment (Wally5077) [inline application/src/main/kotlin/tw/waterballsa/gaas/application/usecases/GetUserUseCase.kt:16]: **static import**

**Review by Wally5077 — CHANGES_REQUESTED** [review]
**Review done.**

**Pull request merged**